fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/expressions/FieldKeyExpression.java

```diff
@@ -313,21 +320,40 @@ public boolean equals(Object o) {
         }
 
         FieldKeyExpression that = (FieldKeyExpression)o;
-        return this.fieldName.equals(that.fieldName) && (this.fanType == that.fanType);
+        return this.fieldName.equals(that.fieldName) &&
+               this.fanType == that.fanType &&
+               this.nullStandin == that.nullStandin;
     }
 
     @Override
     public int hashCode() {
-        return fieldName.hashCode() + fanType.hashCode();
+        return fieldName.hashCode() + fanType.hashCode() + nullStandin.hashCode();
     }
 
     @Override
     public int planHash() {
-        return fieldName.hashCode() + fanType.name().hashCode();
+        int hash = fieldName.hashCode() + fanType.name().hashCode();
+        if (nullStandin == Key.Evaluated.NullStandin.NOT_NULL) {
+            // NULL and NULL_UNIQUE have the same hash, which is also the same as before.
+            hash++;
+        }
+        return hash;
     }
 
     @Override
     public boolean equalsAtomic(AtomKeyExpression other) {
         return equals(other);
     }
+
+    @Override
+    public boolean equivalentForSort(@Nonnull KeyExpression other) {
```

Comment:
I wonder if this should be marked as `EXPERIMENTAL` or something (and maybe the `isPrefixKey` that takes a `forSort`). If, for example, we fixed this long-term this by moving the `NullInterpretation` onto the record's meta-data instead of the index's, we might want to remove this. Maybe that's a long enough project that we handle that by formally deprecating things later.

Reply:
I didn't think about it a lot, but my idea was that this isn't really experimental, even though it further indicates that some changes are needed in this area. I'm happy to change it, though, if there's consensus on how.

Reply:
I guess my thought was that we would eventually want to remove this if we changed how uniqueness on expressions was configured. But I also suppose that there would be a long period where we'd need to support both ways, so maybe just deprecating it then is better.

Reply:
I made it `INTERNAL`, that that seemed a reasonable compromise.